[TAN-4478] Endpoint to copy inputs (2nd attempt)

adessy opened the pull request:
# Changelog
## Technical
- [TAN-4478] Add POST endpoint to copy inputs from a previous phase.

adessy committed:
[TAN-4478] Small Improvement to the Ideas Factory

If phases are specified, the project is set accordingly to ensure consistency.

adessy committed:
[TAN-4478] Add API endpoint to copy ideas (overly simplistic approach)

adessy committed:
[TAN-4520] Add participation stats to common ground results

adessy committed:
[TAN-4478] Move ideas_copy_spec to the correct folder

adessy committed:
[TAN-4478] Extract the logic for copying ideas into a new `Ideas::CopyService`

adessy committed:
Add 'ricecream' gem in development

adessy committed:
[TAN-4478] Add a simple factory for QueJob

adessy committed:
[TAN-4478] Add jobs_trackers table and associated model

Introduced a new jobs_tracked table with a corresponding `Jobs::Tracker` model that will be used to track the progress of `QueJob`s.

adessy committed:
[TAN-4478] Add `TrackableJob` for job progress tracking

adessy committed:
[TAN-4478] Add simple implementation of `jobs` endpoints

Add a simple implementation of `jobs` endpoints to access job progress. It's somewhat redundant with the `.../web_api/v1/background_jobs` endpoints, but hopefully we'll unify them in the future.

adessy committed:
[TAN-4478] Add asynchronous idea copying with job tracking

adessy committed:
Add arm64-darwin-24 platform in Gemfile.lock

adessy committed:
[TAN-4478] Rework the job tracking interface and update Ideas::CopyJob

This commit allows specifying whether tracking should be enabled when queuing the job:

```ruby
ExampleJob.with_tracking.perform_later(*args) # Enqueues with tracking
ExampleJob.with_tracking.perform_now(*args)   # Runs immediately without tracking
ExampleJob.perform_later(*args)               # Enqueues the job without tracking
```

adessy committed:
[TAN-4478] Add ownership to background jobs

Introduced an `owner` association to `jobs_trackers` which can be used to track who queued the job (directly or indirectly).

adessy committed:
[TAN-4478] Refine the `Ideas::CopyService`

Refine the original, simplistic implementation of the idea copy functionality.

adessy committed:
Update .gitignore files

adessy committed:
[TAN-4478] Add documentation comment to `Jobs::TrackableJob#estimate_tracker_total`

adessy committed:
[TAN-4478] Enhance job tracking with completion, context, and error details

Adds new attributes to `Jobs::Tracker`:
* `completed_at` to marks when a job finishes
* `context` and `project_id` to enable filtering of background jobs
* `error_count` to tracks the number of errors for partially successful runs

adessy committed:
[TAN-4478] Refactor owner as an association in serialized job trackers

adessy committed:
[TAN-4478] Fix `owner` association in `Jobs::Tracker` + extra specs

adessy committed:
[TAN-4478] Add filter query params to `GET .../jobs`

adessy committed:
[TAN-4478] Fix `GET web_api/v1/jobs/:id` specs

adessy committed:
[TAN-4478] Rework permissions for accessing jobs

adessy committed:
[TAN-4478] Set the correct creation_phase_id when copying ideas

adessy committed:
[TAN-4478] Add error tracking to `Ideas::CopyJob`

adessy committed:
[TAN-4478] Mark `Ideas::CopyJob` as completed after processing all ideas

adessy committed:
[TAN-4478] Add `dry_run` params `POST .../phases/:to_phase_id/inputs/copy`

Alternatively, this could have been implemented as a `GET .../phases/:to_phase_id/inputs/copy`

adessy committed:
[TAN-4478] Add `error_count` and `completed_at` attributes to serialized job trackers

adessy committed:
[TAN-4478] Add RelatedIdea model to represent relationships between ideas

Introduces a new model to explicitly capture relationships between ideas. Initially, it will be used to mark ideas as copies of other ideas, laying the groundwork for supporting additional relationship types in the future.

adessy committed:
[TAN-4478] Persist relationship to source idea for newly copied ideas

adessy committed:
[TAN-4478] Add `allow_duplicates` parameter to `POST .../phases/:to_phase_id/inputs/copy`

This parameter can be used to specify whether ideas that have already been copied to the target phase should be skipped or not.

adessy committed:
[TAN-4478] Allow total to be 0 in `Jobs::Tracker`

It's okay to queue a job with a `total` (work estimate) of 0, and tracking should not support that. For instance, a job that copies ideas from one phase to another where the source phase has no ideas. While not particularly useful, this should not cause a failure.

Also, in some cases, we expect `total` to be used more as an estimate rather than an exact count. Depending on the job's type and implementation, it's possible for a job with a total of 0 to still perform work at runtime.

adessy committed:
[TAN-4478] Update Tracker associations to handle deletions/nullification properly

adessy committed:
[TAN-4478] Document `Jobs::Tracker`

adessy committed:
[TAN-4478] Change the semantics of `progress` in `Jobs::Tracker`

This updates the meaning of the `progress` attribute in the `Jobs::Tracker` class. Previously:
* `progress + error_count` represented the total work completed
* `progress + error_count <= total`

Now:
* `progress` alone represents the total work completed (both successful and failed)
* `error_count` represents the portion of `progress` that failed
* The relationship is: `error_count <= progress <= total`

adessy committed:
[TAN-4478] Update `Jobs::Tracker` documentation

adessy committed:
[TAN-4478] Adjust `total` when marking `Jobs::Tracker` as complete.

If the job finishes with `progress < total`, we now set `total = progress` when marking it as complete.

adessy committed:
[TAN-4478] Exclude both *originals* and copies when copying ideas with `allow_duplicates` set to `false`

Previously, we were only excluding existing copies.

adessy committed:
Merge pull request #11244 from CitizenLabDotCo/TAN-4478/copy-inputs-endpoint-2-wip

adessy commented:
@kogre, there are a few conflicts and it still needs some testing, but I think you can already start reviewing it. It shouldn’t change much anymore (unless the review calls for it 😛).

adessy committed:
[TAN-4478] Fix tracking in idea copy job

adessy committed:
Merge branch 'TAN-4520/common-ground/progress-endpoint' into TAN-4478/copy-inputs-endpoint-2

# Conflicts:
#	back/app/controllers/web_api/v1/ideas_controller.rb
#	back/db/structure.sql
#	back/spec/models/project_spec.rb

adessy committed:
[TAN-4519] Add page fields to the transient form for Common Ground

kogre commented on back/spec/models/idea_spec.rb:
Totally nitpicking, but not too convinced of this spec. What we're testing here is so declarative in the model, and so many other tests would fail if it were broken, that it's not worth to maintain this imo.

adessy replied:
I get your point. I can remove them if you insist, but I think they’re pretty low-maintenance—especially in the age of LLMs. They also sometimes cover things that aren’t covered by other tests, like `dependent: :destroy`.

kogre commented on back/.gitignore:
```diff
@@ -35,11 +33,4 @@ attractor_output/
 .custom_env
 
 secrets/*
-!secrets/*.sample
-
-# JetBrains settings files
-.idea/
-
-# VSCode settings files
-.vscode/
-.devcontainer/
```
Were these removals deliberate?

adessy replied:
Yes, these were moved to the `.gitignore` at the project root.

kogre commented on back/Gemfile:
```diff
@@ -18,6 +18,7 @@ gem 'csv', '~> 3.3.0'
 gem 'rack-cors', '2.0.0'
 
 group :development, :test do
+  gem 'ricecream', '~> 0.2.1'
```
Handy, TIL

kogre commented on back/app/concerns/jobs/trackable_job.rb:
```diff
@@ -0,0 +1,92 @@
+        @owner = owner
+      end
+
+      delegate :perform_now, to: :@job_class
```
Maybe there are other public methods on the Que or ActiveJob interfaces? Or really only perform_now and perform_later?

adessy replied:
Yes, you're right. I thought about it, then forgot 😬  I think aside from `.set`, we don’t use any other class methods when enqueuing a job. The good thing is, if you do try, it’ll cause a hard fail that should be easy to catch. 

I'll see what I can do at least for `.set`. I don't want to dig in the rabbit hole of `method_missing` if it's not necessary.

adessy replied:
I'll address that one in another PR.

kogre commented on back/app/concerns/jobs/trackable_job.rb:
```diff
@@ -0,0 +1,92 @@
+      raise NotImplementedError
+    end
+
+    def enqueue_child_job(job_class, ...)
+      return job_class.perform_later(...) unless tracked?
+
+      QueJob.transaction do
+        related_job = job_class.perform_later(...)
+        related_job.que_job.update!(data: { root_job_id: root_job_id })
+
+        related_job
+      end
+    end
```
I trust that this works, but as said before part of me doubts whether this mechanism isn't too brittle in 'unknown unknown' ways. I'd rather have us trust on primitives from the job library for this, but I know Que doesn't offer anything. If you judge this to be an absolute necessity for the current feature (copying ideas), then I'm okay with it. Otherwise I'd suggest to simply drop it, and do the copy at once, and wait until we change job libraries at some point.

kogre commented on back/app/controllers/web_api/v1/ideas_controller.rb:
```diff
@@ -303,6 +303,28 @@ def similar_ideas
 
+  def copy
+    dest_phase = Phase.find(params[:phase_id])
+    authorize(dest_phase, :copy_inputs_to_phase?)
```
Shouldn't we also authorize access to the source ideas? Or is that already happening on the controller level or downstream?

adessy replied:
The ideas are "policy_scoped" in the copy job instead. I should probably add some tests around that.

kogre commented on back/app/controllers/web_api/v1/ideas_controller.rb:
```diff
@@ -525,6 +547,18 @@ def not_allowed_update_errors(input)
+  end
+
+  def dry_run?
+    @dry_run ||= params[:dry_run]&.downcase.in? %w[true 1]
```
I think we have `Utils.to_bool` or something that's used in a few places

adessy replied:
Yes, the thing is, the util doesn’t support `nil` as a parameter, and I didn’t want to refactor it in the context of this PR, which is already quite fat. Anyway, I had to rework it because it was buggy 😅

kogre commented on back/app/controllers/web_api/v1/jobs_controller.rb:
```diff
@@ -0,0 +1,37 @@
+      .where(index_params)
+      .order(created_at: :desc)
+      .then { paginate(_1) }
+      .then { policy_scope(_1) }
```
Syntax nitpick: Elegant, but needlessly complex for the next person?

adessy replied:
I actually find it quite readable that way, certainly more so than something like this:
```
job_trackers = paginate(
  policy_scope(
    Jobs::Tracker
      .where(index_params)
      .order(created_at: :desc)
  )
)
```

I’d say it’s on par with:
```
job_trackers = Jobs::Tracker
  .where(index_params)
  .order(created_at: :desc)

job_trackers = policy_scope(job_trackers)
job_trackers = paginate(job_trackers)
```

It’s just a bit more minimalistic, with a more functional feel. Also looking forward to upgrading to the next version of Ruby, where this can be written as:
```
job_trackers = Jobs::Tracker
  .where(index_params)
  .order(created_at: :desc)
  .then { policy_scope(it) }
  .then { paginate(it) }
```

That said, I get that it ultimately comes down to personal preference (kind of like the `unless`/`if` debate, which I’ve never fully understood). Still, I find it pretty frustrating that something like this gets flagged when IMO the complexity of your codebase isn’t coming from things like this.

kogre commented on back/app/jobs/ideas/copy_job.rb:
```diff
@@ -0,0 +1,128 @@
+      summary = Ideas::CopyService.new.copy(batch, dest_phase, current_user)
+      track_progress(summary.count, summary.errors.size)
+
+      if num_remaining_ideas > batch_size
```
Similar to comment above, I wonder whether we really need this and it's worth the extra complexity. How slow is the job copy?

adessy replied:
I’ll address this along with your other comment.

kogre commented on back/app/models/idea.rb:
```diff
@@ -136,6 +136,9 @@ class Idea < ApplicationRecord
   has_many :embeddings_similarities, as: :embeddable, dependent: :destroy
   has_many :authoring_assistance_responses, dependent: :destroy
 
+  has_many :related_idea_associations, class_name: 'RelatedIdea', dependent: :destroy
```
Maybe we could call the model of the join table `IdeaRelation` instead of `RelatedIdea`? Would make this more elegant, and it's also a bit more accurate.

adessy replied:
Good idea! I'll rework that.

kogre commented on back/app/models/related_idea.rb:
```diff
@@ -0,0 +1,48 @@
+# In the future, we may add an explicit +relationship+ attribute (or similar) to this
+# model to support different types of relationships between ideas, such as:
+# "copy", "clone", "duplicate", "generated_from" (e.g., AI-generated), etc.
+class RelatedIdea < ApplicationRecord
```
As said above, small suggestion to change this to `IdeaRelation`. More elegant in association names and more accurate.

kogre commented on back/app/policies/jobs/tracker_policy.rb:
The policy makes sense today, in the assumption that tracking jobs is only an admin/moderator thing. Maybe adding a comment to explain this should be expanded or adapted if we use trackers for other stuff?

adessy replied:
I can try to come up with something that’s not too redundant with the code and tests. I think it’s pretty clear from the code that you need to be at least a moderator to access jobs. I could also add an 'unauthorized' acceptance test.

kogre commented on back/spec/models/jobs/tracker_spec.rb:
Somewhat similar here, some of the most trivial tests don't really add that much imo, if it's just about a single attribute or call in the model of very well tested frameworks. Just adds to the test time and maintenance burden. It's more nuanced here, so okay to keep, just sharing my thoughts.

kogre commented on back/config/routes.rb:
```diff
@@ -77,6 +77,9 @@
       end
 
       resources :background_jobs, only: %i[index]
+      resources :tracked_jobs, only: %i[index show]
```
A bit confused here, isn't it only `jobs` and not `tracked_jobs`, since I only see a `JobsController`?

adessy replied:
Oops, wrong move. I'll remove it.

kogre reviewed (approved):
Very tidy and complete!

Added some comments, most are just thoughts or friendly suggestions. The one larger item is whether we want to support the child jobs/batching mechanism, or not. I'm slightly opposed, unless there is a clear case that the copy really is very slow and we for sure need this.

adessy committed:
[TAN-4519] Set input term to "contribution" for Common Ground

Ideally, it should be "statement", but don't have this term yet.

adessy committed:
Relax the condition used by the `:with_content` scope

An idea is considered to "have no content" only if both the title and the body are missing.

adessy committed:
[TAN-4478] Fix parsing of boolean query params

adessy committed:
Merge pull request #11331 from CitizenLabDotCo/common-ground/refinements-wip

adessy committed:
[TAN-4478] Improve parsing of boolean query params

adessy committed:
Merge branch 'refs/heads/master' into TAN-4478/copy-inputs-endpoint-2

adessy committed:
[TAN-4526] Enable input export for Common Ground

I forgot to set the supports_exports? flag to true for Common Ground, so exports were only partially functional.

adessy commented:
@kogre, trying to pull your comments together:

> The one larger item is whether we want to support the child jobs/batching mechanism or not. I'm slightly opposed, unless there's a clear case where the copy is really slow and we definitely need it.

> I wonder whether we actually need this, and if it’s worth the extra complexity. How slow is the job copy?

But if we flip the question: if the jobs aren’t slow, why bother measuring progress?

I don’t have any definitive data on job durations, but I don’t think it’s unreasonable to assume that copying ideas could take a while, especially if there are images or files attached. I also wanted to build a solution that’s generic enough to be useful in other contexts. This isn’t the first time I’ve needed progress tracking and lacked the proper abstractions.

I agree that it adds some complexity, but I think it’s well-contained and, more importantly, doesn’t surface in the interface.  I also don’t think it would make migrating to another job backend significantly more difficult.

> I trust that this works, but as mentioned before, part of me worries this mechanism might be too brittle in "unknown unknown" ways.

I can’t guarantee there won’t be any issues, but I don’t think this carries more risk than most pieces of code I write. Conceptually, it’s very simple: each `QueJob` holds a reference to the tracker that it updates. In comparison, I think this is much less intrusive than how we store the tenant ID to support multi-tenancy.

> I’d rather rely on primitives from the job library for this, but I know Que doesn’t offer anything. If you believe this is absolutely necessary for the current feature (copying ideas), then I’m okay with it. Otherwise, I’d suggest dropping it and just doing the copy in one go, and waiting until we switch job libraries.

But as you said, they don’t offer anything in that area. We can’t rely on primitives that don’t exist. And as far as I know, the more modern libraries we have in mind (like Solid Queue,  or GoodJob) don’t support progress tracking either. Also, from experience, I know that "at some point" can mean a year, two years, even longer. Even if the current solution isn’t perfect, I think it can still be useful in the meantime.

adessy committed:
[TAN-4478] Remove `tracked_jobs` resource from routes

It was added by mistake; there’s no controller for this resource.

Co-authored-by: Koen Gremmelprez <koen@govocal.com>
Co-authored-by: Adrien Dessy <209268+adessy@users.noreply.github.com>

adessy committed:
[TAN-4478] Add comment to `Jobs::TrackerPolicy::Scope#resolve`

Suggested during PR review.

Co-authored-by: Koen Gremmelprez <koen@govocal.com>
Co-authored-by: Adrien Dessy <209268+adessy@users.noreply.github.com>

adessy committed:
[TAN-4478] Rename `RelatedIdea` to `IdeaRelation`

Suggested during PR review.

Co-authored-by: Koen Gremmelprez <koen@govocal.com>
Co-authored-by: Adrien Dessy <209268+adessy@users.noreply.github.com>

adessy committed:
[TAN-4478] Add `completed` query param to `GET .../jobs`

adessy committed:
[TAN-4478] Add parameter types for `GET .../jobs` endpoint

adessy committed:
Merge branch 'master' into TAN-4478/copy-inputs-endpoint-2

# Conflicts:
#	back/lib/participation_method/common_ground.rb
#	back/spec/factories/ideas.rb
